blog/tech-talk-window-resize-behavior.md

```diff
@@ -0,0 +1,290 @@
+
+In the screenshot, the cyan-tinted frame was the last frame that was being drawn.
+
+The jank to the right of that frame was not tinted cyan. It was tinted in different colors that were still there from previous frames. This indicated that the jank was not coming from the compositor. The compositor was sending the right output.
+
+The compositor is part of Chromium's graphics stack. The following is very simplified, but for the purpose of this blog post you can imagine it like this:
+
+1. The compositor `cc` produces a `CompositorFrame`, which contains draw instructions.
+2. `cc` sends that `CompositorFrame` to the display compositor `viz`.
+3. `viz` then draws the frame and shows it on screen.
+
+Tinting each `CompositorFrame` showed that the compositor produced the right output. So the issue had to be in the display compositor `viz`.
+
+## Locating the relevant `viz` code
+
+From there, I started searching for mentions of Direct3D in the `viz` source code.
+
+_Note: From here on, the post will get a bit more technical and reference source code symbols._
+
+I found that on the ANGLE Direct3D 11 backend, Chromium uses the Windows [DirectComposition](https://learn.microsoft.com/en-us/windows/win32/directcomp/directcomposition-portal) API for drawing the window contents.
+
+DirectComposition differs from all other output devices in Chromium. The DirectComposition `OutputSurface` has the capability `supports_viewporter` ([source link 1](https://source.chromium.org/chromium/chromium/src/+/main:ui/gl/dcomp_presenter.cc;l=144-146;drc=1cffe9643e1e70f1f369bbb28e015d9c5f968546), [source link 2](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/service/display/output_surface.h;l=100-103;drc=c96a878b1cb45f60aac2285ffbdbd6b53dc92415)).
+
+An output surface is a bitmap that can be drawn to, often backed by a GPU texture.
+
+Without `supports_viewporter`, whenever the window size changes, Chromium will create a new output surface matching the new window size. Then it will draw on that surface and show it.
+
+`supports_viewporter` tries to reduce these costly surface allocations. With `supports_viewporter`, Chromium will not allocate a new surface on every resize. Instead, it will allocate a surface that is too large for what we need to draw. Then it will only paint to and show a certain sub-rectangle (the "viewport") of that surface on screen. The other parts of the surface are not supposed to be shown on screen.
+
+This is supposed to make resizing more efficient because all Chromium needs to do is pad the surface to the proper width and height instead of allocating a new surface on every resize. This surface resize logic lives [here](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/service/display/direct_renderer.cc;l=1082;drc=c96a878b1cb45f60aac2285ffbdbd6b53dc92415).
```

Comment:
```suggestion
This is supposed to make resizing more efficient because all Chromium needs to do is pad the surface to the proper width and height instead of allocating a new surface on every resize. This surface resize logic lives in [`direct_renderer.cc`](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/service/display/direct_renderer.cc;l=1082;drc=c96a878b1cb45f60aac2285ffbdbd6b53dc92415).
```

prefer descriptive markdown link titles